actions.go

```diff
@@ -3,10 +3,272 @@ package neo4j
+	err, ids := loadEntityNodes(tx, d.Graph, []int64{d.entityId}, d.Config, findNeighbors, nil)
+	if err != nil {
+		return err
+	}
+	d.dbIds = ids
+	q.actions = append(q.actions, d)
+	return nil
+	// // delete entity and all nodes reachable from this entity,
+	// _, err = tx.Run("MATCH (m) where WHERE ID(m) in $ids DETACH DELETE m,n", map[string]interface{}{"ids": ids})
+	// if err != nil {
+	// 	return err
+	// }
+	// return nil
+}
+
+func (d *DeleteEntity) Run(tx neo4j.Transaction) error {
+	// delete entity and all nodes reachable from this entity,
+	err := d.Queue(tx, nil)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Run("MATCH (m) where WHERE ID(m) in $ids DETACH DELETE m", map[string]interface{}{"ids": d.dbIds})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// used for query
+func mapNeo4jNodes(nodes []neo4jNode) []map[string]interface{} {
+	var result = make([]map[string]interface{}, len(nodes))
+	for index, item := range nodes {
+		result[index] = structs.Map(item)
+	}
+	return result
+}
+
+func mapNeo4jEdges(edges []neo4jEdge) []map[string]interface{} {
+	var result = make([]map[string]interface{}, len(edges))
+	for index, item := range edges {
+		result[index] = structs.Map(item)
+	}
+	return result
+}
+
+func (c *CreateEntity) Queue(tx neo4j.Transaction, q *JobQueue) error {
+	n4jNodes := make([]neo4jNode, 0, c.NumNodes())
+	n4jEdges := make([]neo4jEdge, 0, c.NumEdges())
+	ls.IterateDescendants(c.Node, func(n graph.Node) bool {
+		props := make(map[string]interface{})
+		vars := make(map[string]interface{})
+		n.ForEachProperty(func(s string, in interface{}) bool {
+			prop := c.MakeProperties(n, vars)
+			props[s] = prop
+			return true
+		})
+		node := neo4jNode{
+			labels: []string{c.MakeLabels(n.GetLabels().Slice())},
+			props:  props,
+		}
+		n4jNodes = append(n4jNodes, node)
+		return true
+	}, func(e graph.Edge) ls.EdgeFuncResult {
+		props := make(map[string]interface{})
+		vars := make(map[string]interface{})
+		e.ForEachProperty(func(s string, in interface{}) bool {
+			prop := c.MakeProperties(e, vars)
+			props[s] = prop
+			return true
+		})
+		edge := neo4jEdge{
+			types: c.MakeLabels([]string{e.GetLabel()}),
+			props: props,
+		}
+		n4jEdges = append(n4jEdges, edge)
+		return 0
+	}, false)
+	c.n4jNodes = n4jNodes
+	c.n4jEdges = n4jEdges
+	q.actions = append(q.actions, c)
+	return nil
+}
+
+func (c *CreateEntity) Run(tx neo4j.Transaction) error {
+	if err := c.Queue(tx, nil); err != nil {
+		return err
+	}
+	vars := make(map[string]interface{})
+	query := fmt.Sprintf("CREATE (m %s %s), return m",
+		c.MakeLabels(c.Node.GetLabels().Slice()),
+		c.MakeProperties(c.Node, vars))
+	idrec, err := tx.Run(query, vars)
+	if err != nil {
+		return err
+	}
+	rec, err := idrec.Single()
+	if err != nil {
+		return err
+	}
+	eid := rec.Values[0].(neo4j.Node).Id
+
+	query = "UNWIND $nodeBatch as node UNWIND $edgeBatch as edge MATCH (m) where ID(m)=$eid CREATE (m)-[e]->(n) SET n = node, SET e = edge"
```

Comment:
Does this really work as expected? For Create, I'd expect not to have a MATCH clause, only a CREATE clause. If my understanding of the language semantics is correct, this should create (m)-[e]->(n) as empty nodes and an empty edge first, then set the labels and properties, resulting in duplicates.